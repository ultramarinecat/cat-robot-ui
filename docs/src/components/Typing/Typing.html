<!DOCTYPE html>

<html>
<head>
  <title>Typing.js</title>
  <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  <meta name="viewport" content="width=device-width, target-densitydpi=160dpi, initial-scale=1.0; maximum-scale=1.0; user-scalable=0;">
  <link rel="stylesheet" media="all" href="../../../docco.css" />
</head>
<body>
  <div id="container">
    <div id="background"></div>
    
      <ul id="jump_to">
        <li>
          <a class="large" href="javascript:void(0);">Jump To &hellip;</a>
          <a class="small" href="javascript:void(0);">+</a>
          <div id="jump_wrapper">
          <div id="jump_page_wrapper">
            <div id="jump_page">
              
                
                <a class="source" href="../App/App.html">
                  src/components/App/App.js
                </a>
              
                
                <a class="source" href="../App/ErrorBoundary.html">
                  src/components/App/ErrorBoundary.js
                </a>
              
                
                <a class="source" href="../App/i18n.html">
                  src/components/App/i18n.js
                </a>
              
                
                <a class="source" href="../Auth/Auth.html">
                  src/components/Auth/Auth.js
                </a>
              
                
                <a class="source" href="../Auth/AuthActions.html">
                  src/components/Auth/AuthActions.js
                </a>
              
                
                <a class="source" href="../Auth/AuthReducer.html">
                  src/components/Auth/AuthReducer.js
                </a>
              
                
                <a class="source" href="../Button/Button.html">
                  src/components/Button/Button.js
                </a>
              
                
                <a class="source" href="../Cat/Cat.html">
                  src/components/Cat/Cat.js
                </a>
              
                
                <a class="source" href="../Cat/CatActions.html">
                  src/components/Cat/CatActions.js
                </a>
              
                
                <a class="source" href="../Cat/CatReducer.html">
                  src/components/Cat/CatReducer.js
                </a>
              
                
                <a class="source" href="../Cat/fulltilt.html">
                  src/components/Cat/fulltilt.js
                </a>
              
                
                <a class="source" href="../Connection/Connection.html">
                  src/components/Connection/Connection.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionActions.html">
                  src/components/Connection/ConnectionActions.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionReducer.html">
                  src/components/Connection/ConnectionReducer.js
                </a>
              
                
                <a class="source" href="../Connection/messages.html">
                  src/components/Connection/messages.js
                </a>
              
                
                <a class="source" href="../Intro/Intro.html">
                  src/components/Intro/Intro.js
                </a>
              
                
                <a class="source" href="../Intro/IntroActions.html">
                  src/components/Intro/IntroActions.js
                </a>
              
                
                <a class="source" href="../Particles/Particles.html">
                  src/components/Particles/Particles.js
                </a>
              
                
                <a class="source" href="../Pulse/Pulse.html">
                  src/components/Pulse/Pulse.js
                </a>
              
                
                <a class="source" href="../Router/Router.html">
                  src/components/Router/Router.js
                </a>
              
                
                <a class="source" href="../Sound/CatSound.html">
                  src/components/Sound/CatSound.js
                </a>
              
                
                <a class="source" href="../Sound/Mute.html">
                  src/components/Sound/Mute.js
                </a>
              
                
                <a class="source" href="../Sound/Sound.html">
                  src/components/Sound/Sound.js
                </a>
              
                
                <a class="source" href="../Sound/TurningSound.html">
                  src/components/Sound/TurningSound.js
                </a>
              
                
                <a class="source" href="../Spinner/Spinner.html">
                  src/components/Spinner/Spinner.js
                </a>
              
                
                <a class="source" href="Typing.html">
                  src/components/Typing/Typing.js
                </a>
              
                
                <a class="source" href="../../configureStore.html">
                  src/configureStore.js
                </a>
              
                
                <a class="source" href="../../index.html">
                  src/index.js
                </a>
              
                
                <a class="source" href="../../serviceWorker.html">
                  src/serviceWorker.js
                </a>
              
                
                <a class="source" href="../../setupTests.html">
                  src/setupTests.js
                </a>
              
            </div>
          </div>
        </li>
      </ul>
    
    <ul class="sections">
        
          <li id="title">
              <div class="annotation">
                  <h1>Typing.js</h1>
              </div>
          </li>
        
        
        
        <li id="section-1">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-1">&#182;</a>
              </div>
              <p>Typing effect
Displays text simulating typewritter effect</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>
<span class="hljs-keyword">import</span> React, { memo, useEffect } <span class="hljs-keyword">from</span> <span class="hljs-string">'react'</span>;
<span class="hljs-keyword">import</span> PropTypes <span class="hljs-keyword">from</span> <span class="hljs-string">'prop-types'</span>;

<span class="hljs-keyword">import</span> theaterJS <span class="hljs-keyword">from</span> <span class="hljs-string">'theaterjs'</span>;

<span class="hljs-keyword">import</span> styles <span class="hljs-keyword">from</span> <span class="hljs-string">'./Typing.module.scss'</span>;

<span class="hljs-keyword">const</span> THEATER_SPEED = <span class="hljs-number">1.0</span>;
<span class="hljs-keyword">const</span> THEATER_ACCURACY = <span class="hljs-number">0.8</span>;
<span class="hljs-keyword">const</span> THEATER_WAIT = <span class="hljs-number">500</span>;
<span class="hljs-keyword">const</span> THEATER_ACTOR = <span class="hljs-string">'status'</span>;

<span class="hljs-keyword">let</span> theater;

<span class="hljs-keyword">export</span> <span class="hljs-function"><span class="hljs-keyword">function</span> <span class="hljs-title">Typing</span>(<span class="hljs-params">{ text }</span>) </span>{
  useEffect(<span class="hljs-function"><span class="hljs-params">()</span> =&gt;</span> {
    theater = theaterJS();</pre></div></div>
            
        </li>
        
        
        <li id="section-2">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-2">&#182;</a>
              </div>
              <p>init theater</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>    theater.addActor(
      THEATER_ACTOR,
      {
        <span class="hljs-attr">speed</span>: THEATER_SPEED,
        <span class="hljs-attr">accuracy</span>: THEATER_ACCURACY
      },
      <span class="hljs-string">`.<span class="hljs-subst">${styles.main}</span>`</span>
    );
  }, []);

  useEffect(<span class="hljs-function"><span class="hljs-params">()</span> =&gt;</span> {
    <span class="hljs-keyword">if</span> (theater) {</pre></div></div>
            
        </li>
        
        
        <li id="section-3">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-3">&#182;</a>
              </div>
              <p>type text</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>      theater.addScene(<span class="hljs-string">`<span class="hljs-subst">${THEATER_ACTOR}</span>:<span class="hljs-subst">${text}</span>`</span>, THEATER_WAIT);
    }
  }, [text]);

  <span class="hljs-keyword">return</span> <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-name">div</span> <span class="hljs-attr">className</span>=<span class="hljs-string">{styles.main}</span> <span class="hljs-attr">data-testid</span>=<span class="hljs-string">"typing"</span> /&gt;</span>;
}

Typing.propTypes = {
  text: PropTypes.string.isRequired
};

export default memo(Typing);

</span></pre></div></div>
            
        </li>
        
    </ul>
  </div>
</body>
</html>
