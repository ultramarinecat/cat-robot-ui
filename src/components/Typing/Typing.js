//
// Typing effect
// Displays text simulating typewritter effect
//

import React, { memo, useEffect } from 'react';
import PropTypes from 'prop-types';

import theaterJS from 'theaterjs';

import styles from './Typing.module.scss';

const THEATER_SPEED = 1.0;
const THEATER_ACCURACY = 0.8;
const THEATER_WAIT = 500;
const THEATER_ACTOR = 'status';

let theater;

export function Typing({ text }) {
  useEffect(() => {
    theater = theaterJS();

    // init theater
    theater.addActor(
      THEATER_ACTOR,
      {
        speed: THEATER_SPEED,
        accuracy: THEATER_ACCURACY
      },
      `.${styles.main}`
    );
  }, []);

  useEffect(() => {
    if (theater) {
      // type text
      theater.addScene(`${THEATER_ACTOR}:${text}`, THEATER_WAIT);
    }
  }, [text]);

  return <div className={styles.main} data-testid="typing" />;
}

Typing.propTypes = {
  text: PropTypes.string.isRequired
};

export default memo(Typing);
