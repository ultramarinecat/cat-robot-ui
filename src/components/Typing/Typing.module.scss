@charset 'utf-8';

////
//// Typing styles
////

@import 'base.scss';

.main {
  margin-top: rem(66);
  font-size: rem(25);
  min-height: rem(37);
  @media #{$tablet-plus} {
    margin-top: rem(105);
    font-size: rem(28);
    min-height: rem(42);
  }
}
