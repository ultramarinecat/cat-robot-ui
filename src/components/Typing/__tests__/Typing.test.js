import React from 'react';
import { render } from 'react-testing-library';

import { Typing } from '../Typing.js';

const TYPING = 'typing';

describe('Typing', () => {
  it('should type `text`', () => {
    const text = 'cat';

    const { container, getByTestId } = render(<Typing text={text} />);
    const typing = getByTestId(TYPING);

    expect(typing).toBeInTheDocument();
    expect(typing).toHaveTextContent(text);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should replace previously typed text', () => {
    const initialText = 'cat';

    const { getByTestId, rerender } = render(<Typing text={initialText} />);
    const typing = getByTestId(TYPING);

    expect(typing).toBeInTheDocument();
    expect(typing).toHaveTextContent(initialText);

    const updatedText = 'dog';
    rerender(<Typing text={updatedText} />);

    expect(typing).not.toHaveTextContent(initialText);
    expect(typing).toHaveTextContent(updatedText);
  });
});
