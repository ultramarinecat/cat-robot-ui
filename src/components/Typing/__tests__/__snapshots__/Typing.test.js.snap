// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typing should type \`text\` 1`] = `
<div
  class="main"
  data-testid="typing"
>
  cat
</div>
`;
